pvlib/modelchain.py

```diff
@@ -810,32 +810,102 @@ def run_model(self, times=None, weather=None):
+            also required to compute the missing 'F' or 'H'.
 
         Returns
         -------
         self
 
-        Assigns attributes: times, solar_position, airmass, irradiance,
-        total_irrad, effective_irradiance, weather, temps, aoi,
-        aoi_modifier, spectral_modifier, dc, ac, losses.
+        For conventional weather data, assigns attributes: times,
+        solar_position, airmass, irradiance, total_irrad, effective_irradiance,
+        weather, temps, aoi, aoi_modifier, spectral_modifier, dc, ac, losses.
+        For complete reference device data, only assigns attributes: times, dc,
+        ac, losses.
         """
 
-        self.prepare_inputs(times, weather)
-        self.aoi_model()
-        self.spectral_model()
-        self.effective_irradiance_model()
-        self.temp_model()
+        if weather is not None and 'F' in weather:
+            # Use PV reference device weather information for F
+            self.F = weather['F']
+        else:
+            # Clear any stale value
+            if hasattr(self, 'F'):
+                del self.F
+        if weather is not None and 'H' in weather:
+            # Use PV reference device weather information for H
+            self.H = weather['H']
+        else:
+            # Clear any stale value
+            if hasattr(self, 'H'):
+                del self.H
+        if not hasattr(self, 'F') and not hasattr(self, 'H') or \
+           hasattr(self, 'F') and not hasattr(self, 'H') or \
+           not hasattr(self, 'F') and hasattr(self, 'H'):
+            # Use conventional (meterological) weather information
+            self.prepare_inputs(times, weather)
+            self.aoi_model()
+            self.spectral_model()
+            self.effective_irradiance_model()
+            self.temp_model()
         self.dc_model()
         self.ac_model()
         self.losses_model()
 
         return self
+
+
+def sdm_campanelli(mc):
+    """
+    ModelChain wrapper for Campanelli et al. model, optionally using SAPM for
+    effective irradiance F and/or effecitve temperature H.
+
+    TODO Details
+    """
+
+    if not hasattr(mc, 'F') or not hasattr(mc, 'H'):
+        # SAPM computes F and H so one/both may be used
+        F, H = get_F_H_from_sapm(mc.effective_irradiance, mc.temps,
+                                 mc.system.module_parameters['Aisc'],
+                                 mc.system.module_parameters['irrad_ref'],
+                                 mc.system.module_parameters['temp_ref'])
+        if not hasattr(mc, 'F'):
+            mc.F = F
+        if not hasattr(mc, 'H'):
+            mc.H = H
+
+    # Calculate the system-scaled dc power and assign it to mc.dc
+    mc.dc = mc.system.scale_voltage_current_power(
+        pvsystem.sdm_campanelli(mc.F, mc.H, **mc.system.module_parameters))
+
+    # Return ModelChain object to enable method chaining
+    return mc
+
+
+def get_F_H_from_sapm(effective_irradiance, temps, alpha_sc, irrad_ref,
```

Comment:
same issue with ``get_`` prefix here as in comment on ``get_calcparams_campanelli``. Why did you put this function in ``modelchain.py`` rather than ``pvsystem.py``? Seems to me that it's more general purpose than the ``modelchain`` api.

Reply:
I was thinking that it might make sense to move this function to the *module* ``pvsystem``, but not necessarily into the *class* ``pvsystem.PVSystem``. So, function name changes aside, we'd have ``pvsystem.get_F_H_from_sapm``. One could optionally implement ``pvsystem.PVSystem.get_F_H_from_sapm`` that dispatches work to ``pvsystem.get_F_H_from_sapm``, but that may or may not be of value.